torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -743,6 +757,145 @@ c10::optional<ScheduleHeuristic> tryMerge(
+    unionInto(ab, a);
+    unionInto(ab, b);
+    ab_set->erase(a);
+    ab_set->erase(b);
+    reach_map_.erase(a);
+    reach_map_.erase(b);
+    for (auto& it : reach_map_) {
+      if (it.second->count(a) || it.second->count(b)) {
+        it.second->insert(ab);
+        unionInto(it.first, ab);
+      }
+      it.second->erase(a);
+      it.second->erase(b);
+    }
+  }
+
+ private:
+  //! Collect initial reachable info using
+  //!  a work list algorithm through forward traversal
+  //!  a backward DFS would do the same
+  void computeReachability() {
+    GroupSet completed;
+    GroupSet work_list;
+
+    // Collect source nodes, we are guaranteed
+    //  a source node on a DAG and thus we can
+    //  be sure to make progress
+    std::copy_if(
+        segmented_fusion_->groups().begin(),
+        segmented_fusion_->groups().end(),
+        std::inserter(completed, completed.end()),
+        [](SegmentedGroup* g) { return g->producer_edges.empty(); });
+
+    // completed now only contain source nodes
+    //  they can go backward to nowhere
+    for (auto g : completed) {
+      addConsumersToWorkList(g, work_list);
+    }
+
+    while (!work_list.empty()) {
+      SegmentedGroup* to_update = nullptr;
+      for (auto g : work_list) {
+        if (std::all_of(
+                g->producer_edges.begin(),
+                g->producer_edges.end(),
+                [&completed](SegmentedEdge* e) {
+                  return completed.count(e->from);
+                })) {
+          // filter multi-edges
+          GroupSet producer_groups;
+          for (auto e : g->producer_edges) {
+            producer_groups.insert(e->from);
+          }
+
+          // populate all possible paths
+          // from producer backward, including
+          // the producer
+          for (auto p : producer_groups) {
+            insert(g, p);
+            unionInto(g, p);
+          }
+          to_update = g;
+          break;
+        }
+      }
+      if (to_update) {
+        addConsumersToWorkList(to_update, work_list);
+        work_list.erase(to_update);
+        completed.insert(to_update);
+      } else {
+        TORCH_INTERNAL_ASSERT(false, "unreachable");
+      }
+    }
+  }
+
+  void addConsumersToWorkList(SegmentedGroup* completed, GroupSet& work_list) {
+    for (auto e : completed->consumer_edges) {
+      // A consumer wouldn't have been worked before any of its producer
+      work_list.insert(e->to);
+    }
+  }
+
+  // Compute the union of two reachable maps and store in
+  //  in the entry of to
+  void unionInto(SegmentedGroup* to, SegmentedGroup* from) {
+    auto& from_set = *getSet(from);
+    for (auto g : from_set) {
+      insert(to, g);
+    }
+  }
+
+  //! Utility to access reachable map
+  GroupSetPtr& getSet(SegmentedGroup* key) {
+    auto& ptr = reach_map_[key];
+    if (!ptr) {
+      ptr = std::make_unique<GroupSet>();
+    }
+    return ptr;
+  }
+
+  void insert(SegmentedGroup* key, SegmentedGroup* value) {
```

Comment:
Can you make this function name a bit more specific. insert is such a common container function to see something so simple as `insert(to, g)` just seems a bit off. `insertBackwardReachable` is verbose to the point it may seem unnecessary, but it make the point immediately clear that it's a function inserting something associated with the class we're looking at. I understand what's going on, but there's a lot of very generic sounding functions:
`getSet`
`insert`
`work_list`
`to_update`
I'm not recommending going and changing all of these, insert was just the one that threw me off the most. However, writing code should be like writing a document, and I always encourage names that can get the point across quickly. For example one of Carilli's function name: `currentStreamCaptureStatusMayInitCtx` That's quite verbose, but also leaves little to the imagination.

Reply:
This function also seems like it provides very minimal abstraction, may want to leave it out alltogether.

Reply:
Thanks for the style guideline and sorry for the confusion. I have went through the code and hopefully renamed all the confusing parts. Also removed `insert` method and used inline replacements.